feat: added News post feature based on PagesCMS

bodom0015 opened the pull request:
## Problem
We would like to update the ChiVes UI with a list of recent News Items and/or Blog posts

The content for these posts would be made available by a CMS system

Fixes #204

## Approach
* feat: added a bit to `lib/buildPublicData.js` that will read our `public/content/posts` directory and build up a `posts.json` from the content there
* feat: added `components/Pages/Blog.js`, which will render either BlogList (w/o slug) or BlogPost (w/ slug)
    * added a shared route that detects if slug is present and renders content accordingly :
        * when no slug is provided: `components/Blog/BlogList.js` will render the list of existing Blog posts
        * when slug is provided: `components/Blog/BlogPost.js` will render the contents of a specific Post
* feat: added a section to the homepage to display the 3 most recent blog posts
    * NOTE: if there are less than 3 posts, the homepage text automatically updates to account for this

**PagesCMS: Collection view**
![Screenshot 2025-04-18 at 2 17 58 AM](https://github.com/user-attachments/assets/ba40b94d-6ac4-4053-9dba-9804a4f8c1fc)

**PagesCMS: Post view**
![Screenshot 2025-04-18 at 2 18 06 AM](https://github.com/user-attachments/assets/99f3c083-3e08-4540-9e0e-59e4ca5fa8f1)

**ChiVes: Home Page**
![Screenshot 2025-04-18 at 3 50 20 AM](https://github.com/user-attachments/assets/2d7c9b45-230c-48ba-bd22-1019174ff2cb)

**ChiVes: BlogList view**
![Screenshot 2025-04-18 at 2 50 12 AM](https://github.com/user-attachments/assets/7b848b29-7c8a-4337-8021-fb347b63deb9)

**ChiVes: BlogPost view**
![Screenshot 2025-04-18 at 2 50 01 AM](https://github.com/user-attachments/assets/ade0dd48-1e0b-4294-ac8f-dd37f4f10d8e)

### TODOs
- [x] Images that use a relative URL currently do not work - additional parsing logic might be needed to turn these into absolute URLs
- [x] List 3 (most recent?) news posts on homepage (design TBD)
- [x] styling likely needs work (design TBD)
- [x] Clickable links likely need to be brand green (to match the rest of the UI style?)
- [x] ~~might need additional CSS to pull the Footer to the bottom of the viewport 🤔 not sure why that's different here vs other pages~~

## How to Test
1. Navigate to https://app.pagescms.org/healthyregions/chicago-environment-explorer/204_add_blog/collection/posts
    * You should be prompted to sign in with your GitHub credentials
2. Sign in using your GitHub credentials
    * You should see the PagesCMS UI that allows you to create, view, and edit blog posts
    * You should see that there is an existing post here, titled `This is a test post`
3. Navigate to https://deploy-preview-210--chicago-env-explorer.netlify.app
4. Scroll down to the new "Blog" section
    * You should see a short description of the blog feature here
    * You should see a link to the `/blog` page is offered in the text
    * You should see a listing of the (up to) 3 most recent blog posts here
        * Note that this reuses the same BlogList component as we see below
5. Navigate to https://deploy-preview-210--chicago-env-explorer.netlify.app/blog
    * You should be brought to the BlogList view
        * Note that this reuses the same BlogList component as we saw above
    * You should see a new page here that lists the existing Blog posts that were created using PagesCMS
    * You should see that each post has a `tagline` consisting of the first ~25 words of the post
    * You should see that most GitHub markdown is rendered properly in the tagline (bold/italics work properly, TODO: strikethrough and underline do not render properly)
    * You should see that each post is clickable if you hover over it
6. Click on the test post
    * You should be brought to the BlogPost view
    * You should see the full markdown contents of the post rendered here
    * You should see that inline images are rendered properly
    * You should see a Back button is offered
7. Click the Back button
    * You should be brought back to the BlogList view

mradamcox reviewed (changes requested):
@bodom0015 This is looking good, a few questions/requests though:
1. Why did you decide to put the entire post content into `posts.json`, instead of just the date, title, and a tagline as we were planning? I get it removes the need for an extra fetch on post page load, but I'm worried that eventually we'll just have one really big `posts.json` file that needs to load whenever any individual post is displayed.
2. I can get to the /news page, but clicking through to the post itself doesn't load anything, /news/test-post just shows the main post list.
3. Yeah, let's get the link style to match the rest of the website
4. Interesting about the images. Looks like they get created in the markdown as `/public/img/image.png`, when they should actually be just `/img/image.png` (no `/public` prefix. I have a hunch this can be solved with the PagesCMS config.
5. The footer is not clamped to the bottom on other pages, other pages just have more content (see /community and just zoom the browser out). It would be nice to make that change but not needed for now.

bodom0015 commented:
Hi @mradamcox! Great questions, and I will try to detail my reasoning below. I would be happy to meet and chat about this more in-depth, if it would be helpful 🙏  

1. We build up `posts.json` during the `build:data` phase to parse out the frontmatter + content. Attempting to render the raw markdown file doesn't handle the frontmatter properly (see below). I had looked into a solution using `remark-frontmatter`, but I couldn't figure out how to get this plugin working correctly and this workaround appeared to be a clean solution.
    1. Caching: The `posts.json` is only loaded on the first request to this page - since NewsList and NewPost appear on the same page, we use the same data here. What this means is that if a user views the posts list, selects a post to read, hits the back button to choose more posts, and repeats this process, we only send out a single request to fetch `posts.json`. So this fetch should only happen one time on the first page load, and then we read this same cached posts list to render each post. This also prevents us from needing to parse out the frontmatter on the UI side, which does not seem to have as good of support when using `react-markdown`. Also worth noting that the browser could cache this file, but we currently have disabled caching (likely since Internet Explorer used to be a bit overzealous and cache all responses, including those from APIs) 
    2. **Why not use the same pattern as OEPS?** Our existing pattern from OEPS does not apply, since those markdown files don't contain actual frontmatter - OEPS markdown files simply list the fields in the markdown contents (see [example](https://github.com/GeoDaCenter/opioid-policy-scan/blob/main/data_final/metadata/Access_MOUDs.md?plain=1)). So in the frontend there, they can simply render the markdown from the file and not worry about frontmatter rendering properly.
    3. **Why not use the same pattern as SDOH&Place?** Our existing pattern for SDOH does not apply, since we use a NextJS plugin like [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote) whereas ChiVes is not a NextJS app and cannot follow this same pattern. The underlying markdown files do use the same format, but sadly we can't reuse that technology here without rewriting a good portion of the app into NextJS.

Attempting to render markdown (w/ frontmatter) directly gives us this ugly mess:
![Screenshot 2025-04-16 at 1 09 15 PM](https://github.com/user-attachments/assets/c194c584-45c3-427e-a229-13e008ad2785)

Our caching is currently disabled for `posts.json`, but we could enable this if we're concerned about fetch times/sizes:
![Screenshot_2025-04-16_at_12_59_14_PM](https://github.com/user-attachments/assets/037129dc-e20e-4ae2-952f-dfa49afb4525)

2. Thank you for pointing this out! Sorry, I had forgotten to add the `/:slug?` parameter to the `/news` endpoint - this should be fixed now 

3. Looking at the homepage, I see blue link 🤔 but I thought I remembered @shubhamk008 say that clickable links should be green for ChiVes. I may need to reach out to ask for confirmation here.. maybe I'm misremembering things?

4. I almost wrote some code to "translate" those relative URLs in the markdown, but stopped short of that. If we can't solve it using PagesCMS config, I can give this a try instead

5. Ahhhh thank you, that is good to know. I won't worry about this part for now then :+1:

mradamcox reviewed (changes requested):
Splendid, thanks for this detailed response! All of the considerations make more sense to me now. This is looking really close, though it would be best to merge it only after we have the first post or two in this branch... something better than just the test post. I'll work on getting that content. In the meantime, here are a couple of other small changes:
1. Change "News Posts" header to "Blog"
2. Change link title in nav menu from "News" to "Blog"
3. Change url from /news to /blog
4. Remove margin-bottom on post title in /blog listing
5. Augment the build:data process to generate a tagline from each blog and add this to the entry. I'm thinking the process would be:
    - find the first paragraph of content that doesn't start with `#` (so, avoid making an initial header into the tagline)
    - take the first 25 words, add ellipse ... (if the first 25 words end with a . then only add ..) 
    - Remember that when this is rendered on the frontend, render it as markdown (to capture any italic/bold, etc.
    - There are tons of other potential edge cases here, but let's not worry about them for now.
 6. Display the tagline in the post list between the title and date, text color dark grey (see if there are good examples for this color elsewhere on the site).

bodom0015 commented:
Updated this PR:
* added logic to generate and display the tagline for each post :+1: also supports most basic markdown
* all references/routes to "News" should be changed back to "Blog" :+1:
* improved initial styling, with updated screenshots above :+1: designs still TBD for this view, but can revisit once they are available

mradamcox reviewed (approved):
Fantastic, @bodom0015 everything looks great and I appreciate your attention to detail on this! All of the pieces work well together, and while we'll have style/feature updates and such to make it the future, I think this is a perfect foundation to use for the beginning of the blog. Because this is going to be merged into main and go live right away, and I don't yet have the content for the first blog post, what I'm going to do is delete these test posts and add a basic "Welcome to the chives blog" type of post. That way we'll have a placeholder on the site, and we can unpublish when we have the real first blog content.

Also: We should think about branching/pull request patterns that we want to set up for when people are editing and creating new content. It probably isn't a good plan to have every little edit be saved as a commit on the main branch, which is how the default setup would work. Will think about this more...
